alternator: apply default_time_to_live to update operations

psarna opened the pull request:
CQL tables have a default_time_to_live parameter which allows
specifying a default TTL, which gets automatically applied
for each insert/update. Alternator used to ignore that parameter,
but it might be useful to set it via CQL and have an alternator
table which automatically applies time to live to all updates.

Refs: #9465

psarna commented:
Note: I only tested it semi-manually (aside from the regression tests, that all existing alternator tests continue to pass), by running a modified alternator pytest and then changing the default ttl via cqlsh. In general, it will be a little tricky to test automatically, because first of all it's ttl, so timing based, and secondly we need to access the table from both alternator API and CQL.The changes are quite trivial though, so I wonder how badly we want to have an automatic test case for it - I'm open to suggestions.

eliransin commented:
@psarna I will predict what @nyh first comment is going to be:
Where is the unit test/s for validating this? :slightly_smiling_face:

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1645/

nyh commented:
I think this non-dynamodb feature should be documented somewhere in doc/alternator. If we find it too awkward to document, maybe we didn't do it properly? In particular, why configure this option via CQL making this option harder to explain and justify (and Also, I think, harder to use)? And what happens to the TTL after an UpdateItem? See also my comments in
https://github.com/scylladb/scylladb/issues/9465.
Finally it sounds like this PR should "Fixes", not just "Refs", https://github.com/scylladb/scylladb/issues/9465. If it doesn't Fixes it, we should explain why it doesn't.

psarna commented:
I marked the PR as "Refs", because it only "fixes" our previous incorrect assumption, that setting a default_time_to_live parameter in an alternator table via CQL is a viable workaround for providing limited TTL capability to alternator tables. I think that it would make perfect sense to create another pull request on top of this one, which actually provides an extension to alternator interface and allows setting it without CQL.

I can also try including both this change and the new alternator-specific interface in a single pull request, but if that takes a long time and/or proves to be too hacky, we won't even have the workaround available for users who want the feature.

psarna commented:
Now that I think of it, perhaps an alternator tag (like, `system.scylla.alternator.default_time_to_live` or some other equally unique-enough name) can serve as such configuration. It's already per-table and we can simply add a specialized code path if somebody changes the value of this tag - it simply changes the underlying `default_time_to_live` property in our internal table schema.

psarna commented:
our existing reserved name for the write isolation tag is `system:write_isolation` - with this scheme, `system:default_time_to_live` sounds like a right thing to use

psarna commented:
I just converted the series to draft, because I added experimental, untested changes which introduce a new tag - `system:default_time_to_live`, which changes the value of `default_time_to_live` parameter in the internal schema when changed.
With that, it becomes possible to interact with this setting 100% through alternator interface, bypassing CQL.

Technically we could also just ignore CQL's `default_to_live` parameter and just take the value straight from the tag, but accessing tags is slightly more inconvenient than the existing schema parameter, so perhaps this way is cleaner. I'm fine with both approaches though.

I still need to write tests for this series and test it manually, right now I just post what I have because I'll have a context switch soon.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1663/

nyh commented on alternator/executor.cc:
```diff
@@ -759,6 +759,29 @@ future<executor::request_return_type> executor::tag_resource(client_state& clien
+            auto schema = builder.build();
+
+            auto m = co_await mm.prepare_column_family_update_announcement(schema, false,  std::vector<view_ptr>(), group0_guard.write_timestamp());
+            co_await mm.announce(std::move(m), std::move(group0_guard));
```
Hmm, I guess we can live with this for now (especially since we already have a FIXME here and in db::update_tags) but the right thing isn't to do the change of the schema and the change separately, but rather to prepare one mutation doing both, and then run it together.

psarna replied:
btw, this isn't needed, I can just pass an updated schema pointer straight to db::update_tags and it will be applied

psarna commented on alternator/executor.cc:
```diff
@@ -1143,20 +1143,34 @@ future<executor::request_return_type> executor::update_table(client_state& clien
-    attribute_collector() : collected(attrs_type()->get_keys_type()->as_less_comparator()) { }
+    explicit attribute_collector(gc_clock::duration default_ttl)
+        : collected(attrs_type()->get_keys_type()->as_less_comparator()),
+         _default_ttl{default_ttl}
```
sure, will do

psarna commented on alternator/executor.cc:
```diff
@@ -1143,20 +1143,34 @@ future<executor::request_return_type> executor::update_table(client_state& clien
     }
     void put(const bytes& name, const bytes& val, api::timestamp_type ts) {
-        add(name, atomic_cell::make_live(*bytes_type, ts, val, atomic_cell::collection_member::yes));
+        if (_default_ttl == gc_clock::duration::zero()) {
```
yeah, it crossed my mind but I decided to wait for a review first (in order not to waste time on deduplication in case somebody totally opposes this series), I'll do that in v2

psarna commented:
> Looks good, although, as @eliransin noted, it's not nice that we don't have any tests - not even for error cases. For the success case, we can write a test which will regrettably take one second, which isn't pretty but isn't a complete disaster either.

I'll provide tests soon, I just stalled with them because I wanted to get the design pre-validated. I'll start coding today

> 
> Another thing this patch is missing is documentation. Perhaps the best place to put it in the current organization is the "Alternator-specific API" section of docs/alternator/alternator.md.

Yeah, definitely, I'll provide that in the next iteration
> 
> One thing I'm ambivalent about is the whole idea of using the CQL "default ttl" thing to store the default TTL. On one hand, there is no real reason why we need to support the CQL command to change it, and we could have done this with just the tags without the extra schema change operation. On the other hand, in Scylla this default ttl thing is a field in the **schema**, not just a CQL feature, so arguably we can or even should indeed use this existing schema field. So I don't know which is better.

I wasn't sure either, but two things convince me that using the `default_time_to_live` schema field is acceptable:
 - as you mentioned, it's a schema feature, and we use these schemas in alternator too
 - with the current interface, extracing information out of tags is a little tedious, forcing you to unwrap a string map, looking up a key in this map, deserializing it and then checking what's in there, so it sounds natural to just cache the unpacked value - and the `default_time_to_live` parameter in the schema is a perfect place for that

psarna commented:
v2:
 - deduplicated the code by moving common procedures to helper functions
 - dropped the redundant and incorrect double schema updates in favor of using a single announcement, which updates both the tags and possibly default_time_to_live
 - added test cases to test_ttl.py
 - added a docs paragraph

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1786/

eliransin commented:
Adding a note here so we don't forget to consider before approving this:
@nyh mentioned in the issue (#9465) that odd thing will happen when you update a single attribute or even odder things will
happen when you update a sub attribute.... We could just document those things but it still like too big of a trap for users. I think 
we should take it with @tzach if it is acceptable. IMHO this is too strange and we better go with the trade off approach, meaning, if you use this default TTL you can expect:
1. If read before write operations are enabled then update operation becomes read before write. (at least for now until we see that this feature gets enough traction and it is worth considering optimization of this path)
2. if read before write operations are disabled then update item is not allowed on this table - or any operation which might extend the life of only part of the row.

nyh commented on docs/alternator/alternator.md:
```diff
@@ -129,6 +129,28 @@ isolation policy for a specific table can be overridden by tagging the table
+   on all *new* updated and inserts, and remove them once their expiration period
+   has passed. Note a few important details:
+     - expiration induced by default TTL **does not** produce any Streams events
+     - TTL will only be applied to new updates and inserts, it won't amend existing data in any way
```
Perhaps we should be more explicit here on how this works. I can also send a documentation patch later if you want, but @eliransin is worried that if we describe exactly how this feature works - @tzach and users might complain that this is a bad feature and bad implementation:

The thing is, as this is currently implemented, if UpdateItem (as opposed to PutItem) is used to modify just a single (top-level) attribute, only this attribute will get a a new lease on life, while the rest of the item will expire at its old pre-destined time. The result is that at some point in the future, we'll end up with a partial item, with only the modified attribute and nothing else :-) Presumably this state of affairs makes sense to Scylla users (because Scylla has this feature, and people apparently use it?) but no DynamoDB user will ever guess that this is the case.

We have several alternatives on how to "solve" this issue:
1. Document it (with a bit more detail than you did above) and tell users - "caveat emptor!"
2. Forbid UpdateItem when default-ttl is set
3. Cause UpdateItem to always use read-before-write when default-ttl is set, and always rewrite the entire item.
4. Some sort of combination between 2 and 3 depending on configuration options (a new one or the old write isolation option).

Alternative 1 is easiest, and most "Scylla-like" (Scylla also allows this weird partial-item expiration) but I don't know if @eliransin and @tzach will accept it.

nyh replied:
Another difference from DynamoDB's TTL feature that should probably be mentioned somewhere here is that (like Scylla's TTL), the expiration times here have one second accuracy (the item is deleted within one second of expiring).

nyh commented on docs/alternator/alternator.md:
```diff
@@ -129,6 +129,28 @@ isolation policy for a specific table can be overridden by tagging the table
+   alternator also exposes a way of defining default per-table time to live
+   parameter, expressed in seconds and accessible via a `system:default_time_to_live`
+   tag. The default time to live mechanism will automatically apply TTL
+   on all *new* updated and inserts, and remove them once their expiration period
```
I think we be clearer here that the "TTL" in this mechanism is a relative time - after this number of seconds after the item is written, it gets deleted. This is different from DynamoDB's TTL feature which specifies absolute "expiration time", not a relative TTL.

nyh reviewed (approved):
Looks very good, I'll even forgive you for writing one test that takes 5% of the total Alternator test run time ;-)

However, as @eliransin noted, before committing this we need to decide what to do about the weird semantics of UpdateItem - document them better (I left a comment with some ideas) or somehow "un-Scylla" them (which will lead to ugly code and worse performance).

I also left a couple of other trivial review comments.

psarna commented on alternator/executor.cc:
```diff
@@ -778,7 +791,14 @@ future<executor::request_return_type> executor::untag_resource(client_state& cli
     schema_ptr schema = get_table_from_arn(_proxy, rjson::to_string_view(*arn));
 
     std::map<sstring, sstring> tags_map = get_tags_of_table_or_throw(schema);
+    const bool had_default_ttl = (tags_map.contains("system:default_time_to_live"));
```
indeed, I deliberately always update the schema's ttl in order to have easier access to it, and I didn't use it in the one place I needed it :) will do in v2

psarna replied:
done

psarna commented on alternator/executor.cc:
```diff
@@ -1137,6 +1137,23 @@ future<executor::request_return_type> executor::update_table(client_state& clien
     });
 }
 
+static atomic_cell make_live_cell(const abstract_type& type, api::timestamp_type ts, bytes value,
```
done in v2

psarna commented:
v2:
 - added more paragraphs to docs, further explaining the differences between tag-based TTL and DynamoDB's notion of TTL
 - simplified the code for removing default TTL
 - changed the helper function signature to `const bytes&` to avoid needless copying

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1880/

nyh commented:
> I marked the PR as "Refs", because it only "fixes" our previous incorrect assumption, that setting a default_time_to_live parameter in an alternator table via CQL is a viable workaround for providing limited TTL capability to alternator tables. I think that it would make perfect sense to create another pull request on top of this one, which actually provides an extension to alternator interface and allows setting it without CQL.

Perhaps *now* this PR should "Fixes" instead of "Refs"?

Before merging this PR (which I already "approved" based on my code review), we still need to make a decision on whether this feature, with its "weird" UpdateItem semantics, is acceptable. To remind readers, the "weirdness" that @eliransin pointed to (and I mentioned already in #9465) was that if you create an item, and then update just one attribute, the result is that at some point in the future most of the item will expire but just that one attribute will remain alive for a while longer. We should remember that Scylla's and Cassandra's "default TTL" have exactly the same property, and I'm not aware that anyone complained, so maybe this is an issue that after being documented is perfectly fine to have. Note that a user who only uses PutItem (to replace entire items) and not UpdateItem will not have any problems.

CC @tzach whom we recently surprised by adding other features he wasn't aware of.

mykaul commented:
@tzach - please approve this (see note from @nyh above ).
@nyh - anything else we are waiting for?

tzach commented:
IMHO this fearture intreduce edge cases which are hard to remember and  test, and add little value.
In most case, a user will use either CQL or DynamoDB (Alternator) API - not both.

DynamoDB API users are are migrating from DynamoDB, and does not expect changes in the API.
But maybe this is not an issue, AFAIK, the config table can also be update via Alternator is it the case?
